lib/diamonds.ts

```diff
@@ -0,0 +1,327 @@
+  }
+
+  return false;
+}
+
+// adds unregistered selectors
+export async function addUnregisteredSelectors(
+  diamond: Diamond,
+  contracts: Contract[],
+  exclude: string[] = [],
+): Promise<FacetCut[]> {
+  const diamondFacets: Facet[] = await diamond.facets();
+  const facets = getFacets(contracts);
+  let facetCuts: FacetCut[] = [];
+
+  // if facet selector is unregistered then it should be added to the diamond.
+  for (const facet of facets) {
+    for (const selector of facet.selectors) {
+      const target = facet.target;
+
+      if (
+        target !== diamond.address &&
+        selector.length > 0 &&
+        !selectorExistsInFacets(selector, diamondFacets) &&
+        !exclude.includes(selector)
+      ) {
+        facetCuts.push(
+          printFacetCuts(facet.target, [selector], FacetCutAction.Add),
+        );
+      }
+    }
+  }
+
+  return groupFacetCuts(facetCuts);
+}
+
+// replace registered selectors
+export async function replaceRegisteredSelectors(
+  diamond: Diamond,
+  contracts: Contract[],
+  exclude: string[] = [],
+): Promise<FacetCut[]> {
+  const diamondFacets: Facet[] = await diamond.facets();
+  const facets = getFacets(contracts);
+  let facetCuts: FacetCut[] = [];
+
+  // if a facet selector is registered with a different target address, the target will be replaced
+  for (const facet of facets) {
+    for (const selector of facet.selectors) {
+      const target = facet.target;
+      const oldTarget = await diamond.facetAddress(selector);
+
+      if (
+        target != oldTarget &&
+        target != constants.AddressZero &&
+        target != diamond.address &&
+        selector.length > 0 &&
+        selectorExistsInFacets(selector, diamondFacets) &&
+        !exclude.includes(selector)
+      ) {
+        facetCuts.push(
+          printFacetCuts(target, [selector], FacetCutAction.Replace),
+        );
+      }
+    }
+  }
+
+  return groupFacetCuts(facetCuts);
+}
+
+// removes registered selectors
+export async function removeRegisteredSelectors(
+  diamond: Diamond,
+  contracts: Contract[],
+  exclude: string[] = [],
+): Promise<FacetCut[]> {
+  const diamondFacets: Facet[] = await diamond.facets();
+  const facets = getFacets(contracts);
+  let facetCuts: FacetCut[] = [];
+
+  // if a registered selector is not found in the facets then it should be removed from the diamond
+  for (const diamondFacet of diamondFacets) {
+    for (const selector of diamondFacet.selectors) {
+      const target = diamondFacet.target;
+
+      if (
+        target != constants.AddressZero &&
+        target != diamond.address &&
+        selector.length > 0 &&
+        !selectorExistsInFacets(selector, facets) &&
+        !exclude.includes(selector)
+      ) {
+        facetCuts.push(
+          printFacetCuts(
+            constants.AddressZero,
+            [selector],
+            FacetCutAction.Remove,
+          ),
+        );
+      }
+    }
+  }
+
+  return groupFacetCuts(facetCuts);
+}
```

Comment:
The add/replace/remove function names imply that a transaction is sent, but really they just generate the `FacetCut` formatting.  Should change the name to something like `generateFacetCutsForAdd`.  Can you think of something better?

Reply:
These functions are rather specific, so I think they should throw errors if certain passed data can't be added/removed/replaced.

Reply:
I'd also like a function that takes a list of selectors and targets, and calculates the `FacetCut` diff needed to migrate an existing diamond to the given schema.

Reply:
> The add/replace/remove function names imply that a transaction is sent, but really they just generate the `FacetCut` formatting. Should change the name to something like `generateFacetCutsForAdd`. Can you think of something better?

I asked ChatGPT for help here, how about `addUnregisteredSelectorsToFacetCut`?

Reply:
> I'd also like a function that takes a list of selectors and targets, and calculates the `FacetCut` diff needed to migrate an existing diamond to the given schema.

https://github.com/solidstate-network/solidstate-solidity/pull/158/commits/e702c36dad50b8cea06135110ab87c1d206496e4

Reply:
> These functions are rather specific, so I think they should throw errors if certain passed data can't be added/removed/replaced.

https://github.com/solidstate-network/solidstate-solidity/pull/158/commits/45b2dc44e30d47ece990016ecc65e33dfa97900d